### ISSUE TITLE
^._id in join returns array of all LHS ids, not the specific id for the specific join.

### ISSUE THREAD

**Issue by simen** [sanity-io/gradient#459]
```
*[_type == 'article']{
  "p": *[_id == joinPaths("drafts", ^._id)]
}
```

When I run this, ^._id is not a single id, but an array of all id's on the lhs of the join.

**Comment by erikgrinaker** [comment]
This actually works as expected:

```
*[_type == "movie"]{_id, "m": *[_id == ^._id]}
```

This doesn't:

```
*[_type == "movie"]{_id, "m": *[_id == joinPaths("test", ^._id)]}
```

Panics with:

```
Path expected: [movie_10681 movie_19995 movie_68724 movie_78 movie_926 movie_131631 movie_17654 movie_348 movie_118340 movie_263115 movie_284053 movie_335984 movie_679 movie_126889 movie_157336 movie_286217 movie_315635 movie_353491 movie_70981]
```

**Comment by erikgrinaker** [comment]
I'm a bit stuck on this. I think I've narrowed it down to near the `ConsumeResult()` call in `joinIterator.Next()`. I'm running this query:

```
*[_type == "movie" && _id in ["movie_353491", "movie_118340", "movie_157336"]]{
  _id, "m": *[_type == "movie" && _id == joinPaths("test", ^._id)]
}
```

Debug output gives:

```
search.execution ExecJoin subqueryJoin exec RHS with placeholders expressions.Placeholders{
  "joinValue__id": []interface {}{
    "movie_118340",
    "movie_157336",
    "movie_353491",
  },
}
```

I fixed the panicking joinValue() function call by passing a hard-coded string in the code, to inspect the program flow better.

`joinIterator.Next()` is called only once, thus it needs all three placeholders, so I believe it is correct that it receives them all. The actual panic occurs in `rhsResult.Next()`, which seems to be evaluated once for each result in the inner `*[_type == "movie"]` result (19 matches), in order to filter the entries against `^._id`.

As far as I can tell, the problem is that there is nothing which iterates over the three outer matches, each of which is to be substituted in for `^._id` separately. A possible fix might be to first execute the LHS (outer) and RHS (inner) queries, and then iterate over each entry in the `lhsValues` with the corresponding entry in `ctx.placeholders`, and finally evaluate the `ConsumeResult(rhsResult)` separately for each LHS entry/placeholder pair using a new context containing the single associated placeholder.

Does that seem at all reasonable?

**Comment by atombender** [comment]
I suspect it's a bit more complicated. Expressions like these need to be "exploded" for each LHS, so that if there are N values, `_id == joinPaths("test", ^._id)` becomes `_id == joinPaths("test", V1) || _id == joinPaths("test", V2) || ... || _id == joinPaths("test", VN)`. This is something I've been working on, so I think it's probably something that's best assigned to me.

**Comment by erikgrinaker** [comment]
Yeah, I figured it was a bit naïve. Probably better that you handle it, I'm still finding my way around the query engine.
